### PR TITLE
Added support for multi page TIFFs

### PR DESCRIPTION
**Pull request by Brigadier1-9**
This PR adds basic support for multi page TIFF files by supporting a 'page' property on the options object; this allows selecting of a particular page of a multi page TIFF file.

to do this, just add the page to the options:

```
// selects the 5th page of a multi page TIFF file
var image = sharp(file, {page:4})
```
Note the page counts are zero based.

There is no support for page count retrieval included as, in my use case, I need to retrieve all the pages one at a time anyway, so I am dealing with that by looping though the source tiff until I get an error.

### CODE REVIEW

File: lib/input.js [hunk 1]
```diff
@@ -31,6 +31,13 @@ const _createInputDescriptor = function _createInputDescriptor (input, inputOpti
         throw new Error('Invalid density (1 to 2400) ' + inputOptions.density);
       }
     }
+    // page number
+    if (is.defined(inputOptions.page)) {
+      if (is.integer(inputOptions.page)) {
+        inputDescriptor.page = inputOptions.page;
+      }
+      // todo: check page range?
```
Comment (lovell) [inline lib/input.js]: Yes please, between 0 and [100000](https://github.com/jcupitt/libvips/blob/master/libvips/foreign/tiffload.c#L101).

File: src/common.cc [hunk 2]
```diff
@@ -45,6 +45,10 @@ namespace sharp {
       descriptor->rawWidth = AttrTo<uint32_t>(input, "rawWidth");
       descriptor->rawHeight = AttrTo<uint32_t>(input, "rawHeight");
     }
+    // Page
+    if (HasAttr(input, "page")) {
+        descriptor->page = AttrTo<uint32_t>(input, "page");
```
Comment (lovell) [inline src/common.cc]: indentation?

File: src/common.cc [hunk 3]
```diff
@@ -223,6 +227,10 @@ namespace sharp {
           if (imageType == ImageType::MAGICK) {
             option->set("density", std::to_string(descriptor->density).data());
           }
+          if (imageType == ImageType::TIFF) {
+            //printf("It is a TIFF File!!  %s", std::to_string(descriptor->page).data());
```
Comment (lovell) [inline src/common.cc]: stray comment

File: src/common.h [hunk 4]
```diff
@@ -38,6 +38,7 @@ namespace sharp {
     int rawChannels;
     int rawWidth;
     int rawHeight;
+    int page;
```
Comment (lovell) [inline src/common.h]: I think we need to set this to a default of `0`.

**Comment by lovell** [comment]
Thanks for this useful addition, I've added a few comments in line.

Please can you fix the failing tests and add additional tests to cover the new logic added, then rebase+squash into a single commit.

**Comment by Brigadier1-9** [comment]
On it. Thanks.

**Comment by lovell** [comment]
@Brigadier1-9 Do you need any help with this?

**Comment by dstringvc** [comment]
What is the status of this PR? I have an urgent need to convert multi-page TIFF images. Is there anything I can do to help?  Thanks!

**Comment by Brigadier1-9** [comment]
It's kind of stalled; my fault I'm afraid. I have a new version in place (after a refactor of sharp) which I have used to pull about 100,000 images out of a bunch of multi page tiffs with no issues, so I am fairly confident it works. I will try to submit a new PR in the morning and close this one at that point.
Sorry for the delay!

**Comment by dstringvc** [comment]
Any updates on this PR? Thank you!

**Comment by rpachos** [comment]
Hey all, any idea if this PR will move forward? Currently using sharp and discovering a need for this functionality.

**Comment by lovell** [comment]
Hello, I'm going to close this PR as it has drifted too far behind the master branch but please feel free to open a new PR with a similar change if this is still required.

**Pull request closed without merging**